whitepapers/0009_generic_relaying.md

```diff
@@ -0,0 +1,143 @@
+need to be built to specifically parse a particular protocol's payloads. Also incentivizing the "relayer" would need to
+be done at the application level which increases the implementation and operations effort for people building on top of
+Wormhole.
+
+## Goals
+
+There needs to be a higher level protocol on top of wormhole that encapsulates the application payload and adds a
+delivery-specific protocol frame including intended recipient and incentives.
+
+* Provide a simple interface to use a universal delivery layer (at least once)
+* Allow smart contracts to implement the `WormholeReceiver` interface to start receiving messages
+* Gas-effective incentivization mechanism for message delivery
+
+## Non-Goals
+
+* Support delivery on Solana and chains that can't use a generic receiver interface
+* Delivery-confirmation
+* Exactly-once delivery
+* Support fee coins other than the native currency of the chain
+
+## Overview
+
+On each chain of the delivery network there will be a delivery bridge endpoint program.
+
+These programs will manage authorization of payloads (emitter filtering) and publishing of messages.
+
+The protocol wraps the emitted message in a frame encoding the intended recipient of the protocol and delivery fee
+attached. Receivers of Wormhole messages are expected to implement a receiver interface which is universal and the
+off-chain relayers know how to interact with. This interface will interact with the delivery bridge endpoint for
+verification of the VAA and fee accounting.
+
+Fees are collected on the sending side and collected fees are tracked on the receiving side (per relayer). The relayer
+can then claim fees in batch, requesting a payout on the receiving side and claiming them using the produced VAA on the
+receiving side.
+
+## Detailed Design
+
+A proxy contract on each supported chain allows programs to emit wormhole messages with delivery information,
+essentially wrapping it in a relayer protocol frame.
+
+Relayers (running a piece of software supporting this protocol) will watch for messages emitted by the proxy contracts
+and parse the frame. Then they will try to submit it to the target chain using the receiver interface. They will make
+sure that the attached delivery reward/fee is sufficient to cover their cost.
+
+Ideally senders will have an API to estimate gas costs of message delivery using a universal API that simulates the
+delivery of the produced VAA on the target chain (estimating gas). This will require a more complex simulation logic
+that can skip signature verification in the contract (replacing it with a stub; here we need to make sure to still
+account for verification gas which is almost constant) as otherwise, due to the lack of a valid VAA already existing,
+the message would not execute.
+
+When a message is published the sender can attach a fee in a coin. This coin is going to be held in escrow by the proxy
```

Comment:
Why not just pay the fee in target chain's asset on successful submission of the VAA? This would mean instant payment on submission.